builtin/providers/aws/resource_aws_instance_test.go

```diff
@@ -623,6 +625,76 @@ func TestInstanceTenancySchema(t *testing.T) {
 	}
 }
 
+func TestAWSInstanceValidate(t *testing.T) {
+	r := resourceAwsInstance()
+	cases := map[string]struct {
+		Config           map[string]interface{}
+		Vars             map[string]ast.Variable
+		ExpectedErrors   []error
+		ExpectedWarnings []string
+	}{
+		"errors when subnet_id used with security_groups": {
+			Config: map[string]interface{}{
+				"ami":             "ami-abc123",
+				"instance_type":   "t2.micro",
+				"subnet_id":       "subnet-abc123",
+				"security_groups": []interface{}{"my-sg-name"},
+			},
+			ExpectedErrors: []error{errMsgSecurityGroupsNamesWithSubnetID},
+		},
+		"errors when vpc_security_group_ids used without subnet_id": {
+			Config: map[string]interface{}{
+				"ami":                    "ami-abc123",
+				"instance_type":          "t2.micro",
+				"vpc_security_group_ids": []interface{}{"sg-abc123"},
+			},
+			ExpectedErrors: []error{errMsgSecurityGroupIDSWithoutSubnetID},
+		},
+		"does not error in a VPC/Classic optional config setup": {
+			Config: map[string]interface{}{
+				"ami":                    "ami-abc123",
+				"instance_type":          "t2.micro",
+				"subnet_id":              "${var.subnet_id}",
+				"security_groups":        []interface{}{`${split(",", var.sg_names)}`},
+				"vpc_security_group_ids": []interface{}{`${split(",", var.vpc_sg_ids)}`},
+			},
+			Vars: map[string]ast.Variable{
+				"var.subnet_id": ast.Variable{
+					Value: config.UnknownVariableValue,
+					Type:  ast.TypeString,
+				},
+				"var.sg_names": ast.Variable{
+					Value: config.UnknownVariableValue,
+					Type:  ast.TypeString,
+				},
+				"var.vpc_sg_ids": ast.Variable{
+					Value: config.UnknownVariableValue,
+					Type:  ast.TypeString,
+				},
+			},
+		},
```

Comment:
@mitchellh this is the unit test expressing a legitimate use case involving this "unknown" concept. The ValidateFunc implementation needs to be aware of this possibility and handle it correctly.